### PR TITLE
Implement rename option of files

### PR DESCRIPTION
**Pull request by AndriiSameliuk**
Implement rename files option.
image

### CODE REVIEW

**Comment by AndriiSameliuk** [comment]
@dkalinin PTAL.

File: app/View/Component/FileItem/index.js [hunk 1]
```diff
@@ -91,4 +95,29 @@ export class FileItem extends Component {
...
+   * Creates input and button for renaming file.
+   * @param {Event} event
+   * @private
+   */
+  _onDoubleClick(event) {
+    this.div.innerHTML = ``;
+    const text = this.fileName.textContent;
+    const input = new Input(this.div, 'text', 'renaming-input');
+    input.placeholder = text;
+    const button = new Button(this.div, 'Rename');
```
Comment (dkalinin) [inline app/View/Component/FileItem/index.js]: Can be moved to _render if property "isRenaming" created.

File: app/View/Component/FileItem/index.js [hunk 2]
```diff
@@ -91,4 +95,29 @@ export class FileItem extends Component {
...
+  /**
+   * Creates input and button for renaming file.
+   * @param {Event} event
+   * @private
+   */
+  _onDoubleClick(event) {
+    this.div.innerHTML = ``;
+    const text = this.fileName.textContent;
+    const input = new Input(this.div, 'text', 'renaming-input');
+    input.placeholder = text;
```
Comment (dkalinin) [inline app/View/Component/FileItem/index.js]: Should be value.

File: app/Store/Actions/index.js [hunk 3]
```diff
@@ -44,4 +44,17 @@ export default {
...
+
+  renameFile(context, file) {
+    const promise = FileService.renameFile(file.id, file.name);
+    promise
+        .then(() => {
+          context.commit('renameFile', file);
+          context.dispatch('fetchFiles');
+        })
+        .catch(() => {
+        })
```
Comment (dkalinin) [inline app/Store/Actions/index.js]: Please handle server error response.

**Comment by AndriiSameliuk** [comment]
@dkalinin, fixed, PTAL.

File: app/Store/Actions/index.js [hunk 4]
```diff
@@ -45,4 +45,23 @@ export default {
...
+  renameFile(context, fileId, fileName) {
+    context.commit('setFilesLoadingStatus', true);
+    const promise = FileService.renameFile(fileId, fileName);
+    promise
+        .then(() => {
+          context.commit('renameFile', fileId, fileName);
+          context.dispatch('fetchFiles');
+        })
+        .catch(() => {
+          context.commit('setExistFileNameStatus', true);
```
Comment (dkalinin) [inline app/Store/Actions/index.js]: Let's have a field with server error text.

File: app/Store/Mutations/index.js [hunk 5]
```diff
@@ -25,4 +24,19 @@ export default {
     }
     state.files = stateFiles;
   },
+  renameFile(state, fileId, fileName) {
+    const files = state.files;
+    const stateFile = files.filter((file) => file.id === fileId);
+
+    files.forEach((file) => {
+      if (file.id === stateFile[0].id) {
```
Comment (dkalinin) [inline app/Store/Mutations/index.js]: Please check [JS Array Doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array)

File: app/Store/Mutations/index.js [hunk 6]
```diff
@@ -25,4 +24,19 @@ export default {
     }
     state.files = stateFiles;
   },
+  renameFile(state, fileId, fileName) {
+    const files = state.files;
+    const stateFile = files.filter((file) => file.id === fileId);
```
Comment (dkalinin) [inline app/Store/Mutations/index.js]: Please rename stateFile.

File: app/Store/Mutations/index.js [hunk 7]
```diff
@@ -25,4 +24,19 @@ export default {
     }
     state.files = stateFiles;
   },
+  renameFile(state, fileId, fileName) {
```
Comment (dkalinin) [inline app/Store/Mutations/index.js]: fileName => newName

File: app/View/Component/FileItem/index.js [hunk 8]
```diff
@@ -40,10 +41,41 @@ export class FileItem extends Component {
    * Renders spans and buttons.
    */
   _render() {
-    const fileName = document.createElement('span');
-    fileName.textContent = this.name;
-    this.div.appendChild(fileName);
-    this.button = new Button(this.div, 'Delete');
+    this.div.innerHTML = ``;
+    if (!!!this.isRenaming) {
```
Comment (dkalinin) [inline app/View/Component/FileItem/index.js]: One `!` is enough.

File: app/View/Component/FileItem/index.js [hunk 9]
```diff
@@ -40,10 +41,41 @@ export class FileItem extends Component {
...
+    } else {
+      const text = this.fileName.textContent;
+      const input = new Input(this.div, 'text', 'renaming-input');
+      input.value = text;
+      this.renameButton = new Button(this.div, 'Rename');
+
+      this.renameButton.onClick((event) => {
+        event.preventDefault();
+        event.stopPropagation();
+        this.renameButton.disabled = true;
```
Comment (dkalinin) [inline app/View/Component/FileItem/index.js]: Use disabledRenameButton setter.

File: app/View/Component/FileList/index.js [hunk 10]
```diff
@@ -42,15 +44,32 @@ export class FileList extends StoreAwareComponent {
...
+          this.dispatch('renameFile', fileId, name);
+        });
       });
     }
   }
 
+  /**
+   * Checks whether server sends error.
+   */
+  isFileNameExist() {
```
Comment (dkalinin) [inline app/View/Component/FileList/index.js]: Please rename.

File: app/View/Component/FileList/index.js [hunk 11]
```diff
@@ -42,15 +44,32 @@ export class FileList extends StoreAwareComponent {
         fileItem.onFolderDrop = ((files, folderId) => {
           this.dispatch('addItems', files, folderId);
         });
+
+        fileItem.onFileRename = ((fileId, name) => {
+          this.fileItem = fileItem;
```
Comment (dkalinin) [inline app/View/Component/FileList/index.js]: renamedFileItem

File: app/View/Component/FileList/index.js [hunk 12]
```diff
@@ -42,15 +44,32 @@ export class FileList extends StoreAwareComponent {
...
   }
 
+  /**
+   * Checks whether server sends error.
+   */
+  isFileNameExist() {
+    if (this.state.isFileNameExist === true) {
+      this.span.textContent = '';
+      this.fileItem.disabledRenameButton = false;
+      this.fileItem.addErrorSpan('Name is exist.');
```
Comment (dkalinin) [inline app/View/Component/FileList/index.js]: 'Name is exist.' => Name exists.

File: app/Store/Mutations/index.js [hunk 13]
```diff
@@ -19,10 +18,36 @@ export default {
...
     state.files = stateFiles;
   },
+  renameFile(state, fileId, newFileName) {
+    const files = state.files;
+
+    const renamedFile = files.find((file) => {
+      if (file.id === fileId) {
+        return true;
+      }
+    });
```
Comment (dkalinin) [inline app/Store/Mutations/index.js]: `files.find((file) => file.id === fileId)`

File: app/Store/Mutations/index.js [hunk 14]
```diff
@@ -19,10 +18,36 @@ export default {
...
+
+    files.find((file) => {
+      if (file.id === renamedFile.id) {
```
Comment (dkalinin) [inline app/Store/Mutations/index.js]: Why do we search for it again?

File: app/View/Component/FileList/index.js [hunk 15]
```diff
@@ -12,7 +12,12 @@ export class FileList extends DraggableComponent {
    */
   constructor(root) {
     super(root);
-    this.store.events.subscribe('stateChange', () => this.render());
+    this.store.events.subscribe('stateChange.files', () => this.render());
+    this.store.events.subscribe('stateChange.isFilesLoading', () => this.render());
+    this.store.events.subscribe('stateChange.renamedServerError', (errorMessage) =>
```
Comment (dkalinin) [inline app/View/Component/FileList/index.js]: renamedServerError => renamingServerError

File: app/View/Component/Button/index.js [hunk 16]
```diff
@@ -13,19 +13,19 @@ export class Button extends Component {
   constructor(root, name) {
     super(root);
     this._textContext = name;
-    this.initRender();
+    this._render();
```
Comment (dkalinin) [inline app/View/Component/Button/index.js]: Please make such refactoring which involves a whole application in a separate branch in the future.

File: app/View/Component/FileList/index.js [hunk 17]
```diff
@@ -41,10 +46,44 @@ export class FileList extends DraggableComponent {
...
+  }
+
+  /**
+   * Checks fileRenamingStatus.
+   */
+  checkRenamingContinue() {
+    if (this.state.filesRenamingStatus === true) {
+      this.fileList.forEach((fileItem) => {
+        if (fileItem.id !== this.renamedFileItem._id) {
+          fileItem.isRenameContinue = true;
```
Comment (dkalinin) [inline app/View/Component/FileList/index.js]: What if I double click two items in a list and start renaming?

File: app/Store/Actions/index.js [hunk 18]
```diff
@@ -45,4 +45,28 @@ export default {
 
         });
   },
+
+
+  setFilesRenamingStatus(context, value) {
+    context.commit('setFilesRenamingStatus', value);
+  },
```
[the remark this answers is not in the document]
Comment (AndriiSameliuk) [reply app/Store/Actions/index.js]: What should I do in this case?

File: app/View/Component/FileList/index.js [hunk 19]
```diff
@@ -29,22 +32,39 @@ export class FileList extends DraggableComponent {
       this.root.appendChild(this._ul);
       this._addEventHandler();
 
-
+      this.fileList = [];
```
Comment (dkalinin) [inline app/View/Component/FileList/index.js]: Please remove.

**Pull request merged**